tests/components/vizio/test_media_player.py

```diff
@@ -311,3 +415,89 @@
 ) -> None:
     """Test device becomes unavailable after being available."""
     await _test_update_availability_switch(hass, True, None, caplog)
+
+
+async def test_setup_with_apps(
+    hass: HomeAssistantType,
+    vizio_connect: pytest.fixture,
+    vizio_update_with_apps: pytest.fixture,
+    caplog: pytest.fixture,
+) -> None:
+    """Test device setup with apps."""
+    await _test_setup_with_apps(hass, MOCK_USER_VALID_TV_CONFIG, CURRENT_APP)
+    await _test_service(
+        hass,
+        "launch_app",
+        SERVICE_SELECT_SOURCE,
+        {ATTR_INPUT_SOURCE: CURRENT_APP},
+        CURRENT_APP,
+    )
+
+
+async def test_setup_with_apps_include(
+    hass: HomeAssistantType,
+    vizio_connect: pytest.fixture,
+    vizio_update_with_apps: pytest.fixture,
+    caplog: pytest.fixture,
+) -> None:
+    """Test device setup with apps and apps["include"] in config."""
+    await _test_setup_with_apps(hass, MOCK_TV_WITH_INCLUDE_CONFIG, CURRENT_APP)
+
+
+async def test_setup_with_apps_exclude(
+    hass: HomeAssistantType,
+    vizio_connect: pytest.fixture,
+    vizio_update_with_apps: pytest.fixture,
+    caplog: pytest.fixture,
+) -> None:
+    """Test device setup with apps and apps["exclude"] in config."""
+    await _test_setup_with_apps(hass, MOCK_TV_WITH_EXCLUDE_CONFIG, CURRENT_APP)
+
+
+async def test_setup_with_apps_additional_apps_config(
+    hass: HomeAssistantType,
+    vizio_connect: pytest.fixture,
+    vizio_update_with_apps: pytest.fixture,
+    caplog: pytest.fixture,
+) -> None:
+    """Test device setup with apps and apps["additional_configs"] in config."""
+    await _test_setup_with_apps(hass, MOCK_TV_WITH_ADDITIONAL_APPS_CONFIG, UNKNOWN_APP)
+
+    await _test_service(
+        hass,
+        "launch_app",
+        SERVICE_SELECT_SOURCE,
+        {ATTR_INPUT_SOURCE: CURRENT_APP},
+        CURRENT_APP,
+    )
+    await _test_service(
+        hass,
+        "launch_app_config",
+        SERVICE_SELECT_SOURCE,
+        {ATTR_INPUT_SOURCE: CUSTOM_APP_NAME},
+        **CUSTOM_CONFIG,
+    )
+
+    # Test that invalid app does nothing
+    with patch(
+        "homeassistant.components.vizio.media_player.VizioAsync.launch_app"
+    ) as service_call1, patch(
+        "homeassistant.components.vizio.media_player.VizioAsync.launch_app_config"
+    ) as service_call2:
+        await hass.services.async_call(
+            MP_DOMAIN,
+            SERVICE_SELECT_SOURCE,
+            service_data={ATTR_ENTITY_ID: ENTITY_ID, ATTR_INPUT_SOURCE: "_"},
+            blocking=True,
+        )
+        assert not service_call1.called
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5130873)